primitives/src/merkle_tree/internal.rs

```diff
@@ -410,67 +440,39 @@ where
-                    let branch = traversal_path[height - 1];
-                    if !children.iter().zip(proof_children.iter()).enumerate().all(
-                        |(index, (child, proof_child))| {
-                            index == branch
-                                || (matches!(**child, MerkleNode::Empty)
-                                    && matches!(**proof_child, MerkleNode::Empty))
-                                || child.value() == proof_child.value()
-                        },
-                    ) {
-                        Err(PrimitivesError::ParameterError(format!(
-                            "Invalid proof. Sibling differs at height {}",
-                            height
-                        )))
-                    } else {
-                        children[branch].remember_internal::<H, Arity>(
-                            height - 1,
-                            traversal_path,
-                            path_values,
-                            proof,
-                        )
-                    }
-                },
-                MerkleNode::ForgettenSubtree { value: _ } => {
-                    *self = MerkleNode::Branch {
-                        value: path_values[height],
-                        children: {
-                            let mut children = proof_children.clone();
-                            children[traversal_path[height - 1]].remember_internal::<H, Arity>(
-                                height - 1,
-                                traversal_path,
-                                path_values,
-                                proof,
-                            )?;
-                            children
-                        },
-                    };
-                    Ok(())
-                },
-                MerkleNode::Empty => Err(PrimitivesError::ParameterError(
-                    "Invalid proof. Given location is supposed to be empty.".to_string(),
-                )),
-                MerkleNode::Leaf {
-                    value: _,
-                    pos: _,
-                    elem: _,
-                } => Err(PrimitivesError::ParameterError(
-                    "Given position is already occupied".to_string(),
-                )),
-            }
-        } else {
-            Err(PrimitivesError::ParameterError("Invalid proof".to_string()))
+
+        match (&mut *self, &proof[height]) {
+            (Self::ForgettenSubtree { .. }, Self::Branch { children, .. }) => {
+                // Recurse into the appropriate sub-tree to remember the rest of the path.
+                let mut children = children.clone();
+                children[traversal_path[height - 1]].remember_internal::<H, Arity>(
+                    height - 1,
+                    traversal_path,
+                    path_values,
+                    proof,
+                )?;
+                // Compute new value and remember `*self`.
+                *self = Self::Branch {
```

Comment:
I'm thinking about the necessity of recomputing the internal node value. `Remember` function shouldn't change the internal value, right?

Reply:
Oh of course. I was thinking I can't use the value from the proof because they're all zeroed, but forgot I could just get it from `self`. Good catch, will fix